presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -175,33 +179,22 @@ public synchronized void process(Iterable<QueryExecution> queries)
         }
         clusterMemoryUsageBytes.set(totalBytes);
 
-        if (killOnOutOfMemory) {
-            boolean shouldKillQuery = nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) > 0 && outOfMemory;
-            boolean lastKilledQueryIsGone = (lastKilledQuery == null);
-
-            if (!lastKilledQueryIsGone) {
-                ClusterMemoryPool generalPool = pools.get(GENERAL_POOL);
-                if (generalPool != null) {
-                    lastKilledQueryIsGone = generalPool.getQueryMemoryReservations().containsKey(lastKilledQuery);
-                }
-            }
-
-            if (shouldKillQuery && lastKilledQueryIsGone && !queryKilled) {
-                // Kill the biggest query in the general pool
-                QueryExecution biggestQuery = null;
-                long maxMemory = -1;
-                for (QueryExecution query : queries) {
-                    long bytesUsed = query.getTotalMemoryReservation();
-                    if (bytesUsed > maxMemory && query.getMemoryPool().getId().equals(GENERAL_POOL)) {
-                        biggestQuery = query;
-                        maxMemory = bytesUsed;
-                    }
-                }
-                if (biggestQuery != null) {
-                    biggestQuery.fail(new PrestoException(CLUSTER_OUT_OF_MEMORY, "The cluster is out of memory, and your query was killed. Please try again in a few minutes."));
-                    queriesKilledDueToOutOfMemory.incrementAndGet();
-                    lastKilledQuery = biggestQuery.getQueryId();
-                }
+        if (outOfMemory && !queryKilled && nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) > 0 && isLastKilledQueryGone()) {
+            List<QueryMemoryInfo> queryMemoryInfoList = Streams.stream(queries)
+                    .map(query -> new QueryMemoryInfo(query.getQueryId(), query.getMemoryPool().getId(), query.getTotalMemoryReservation()))
+                    .collect(toImmutableList());
+            List<MemoryInfo> nodeMemoryInfos = nodes.values().stream()
+                    .map(RemoteNodeMemory::getInfo)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(toImmutableList());
+            Optional<QueryId> chosenQueryId = lowMemoryKiller.chooseQueryToKill(queryMemoryInfoList, nodeMemoryInfos);
+            if (chosenQueryId.isPresent()) {
+                QueryExecution chosenQuery = Streams.stream(queries).filter(query -> chosenQueryId.get().equals(query.getQueryId())).collect(onlyElement());
+                chosenQuery.fail(new PrestoException(CLUSTER_OUT_OF_MEMORY, "The cluster is out of memory, and your query was killed. Please try again in a few minutes."));
```

Comment:
I realize this is existing, but I dislike "your" in the error message, which is different than other error messages. Let's change this to

*"Query killed because the cluster is out of memory. Please try again in a few minutes."*